gdnative-core/src/nativescript/user_data.rs

```diff
@@ -732,3 +746,76 @@ where
         Ok(op(&Default::default()))
     }
 }
+
+/// Special user-data wrapper intended for objects that can only be used once. Only
+/// implements `MapOwned`.
+pub struct Once<T>(Arc<atomic_take::AtomicTake<T>>);
+
+impl<T> Clone for Once<T> {
+    #[inline]
+    fn clone(&self) -> Self {
+        Once(Arc::clone(&self.0))
+    }
+}
+
+unsafe impl<T> UserData for Once<T>
+where
+    T: NativeClass + Send,
+{
+    type Target = T;
+
+    #[inline]
+    fn new(val: Self::Target) -> Self {
+        Once(Arc::new(atomic_take::AtomicTake::new(val)))
+    }
+
+    #[inline]
+    fn into_user_data(self) -> *const libc::c_void {
+        Arc::into_raw(self.0) as *const _
+    }
+
+    #[inline]
+    unsafe fn consume_user_data_unchecked(ptr: *const libc::c_void) -> Self {
+        Once(Arc::from_raw(ptr as *const _))
+    }
+
+    #[inline]
+    unsafe fn clone_from_user_data_unchecked(ptr: *const libc::c_void) -> Self {
+        let borrowed = Arc::from_raw(ptr as *const _);
+        let arc = Arc::clone(&borrowed);
+        mem::forget(borrowed);
+        Once(arc)
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Default)]
+pub struct ValueTaken;
+
+impl std::error::Error for ValueTaken {}
+impl Display for ValueTaken {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "this object can only be used once")
```

Comment:
Capitalize the first char for consistency with the other `Display` impls in the file. Also, the text alludes to the object having been used already. Perhaps be more direct? "This object has already been used once."

Reply:
Changed to `this object has already been used once`. According to the Rust API Guidelines, error messages should actually be lower case without trailing punctuation: https://rust-lang.github.io/api-guidelines/interoperability.html#error-types-are-meaningful-and-well-behaved-c-good-err

We can change the other `Display` impls in another PR.